broker/src/test/java/io/camunda/zeebe/broker/transport/partitionapi/InterPartitionCommandReceiverTest.java

```diff
@@ -0,0 +1,250 @@
+            commandBytes);
+
+    final var receiver =
+        new InterPartitionCommandReceiver(
+            receiverBrokerId, receiverPartitionId, communicationService, logstream);
+
+    scheduler.submitActor(receiver);
+    scheduler.workUntilDone();
+
+    // when
+    getHandlerForPartition(communicationService, receiverPartitionId).accept(sentMessage);
+    scheduler.workUntilDone();
+
+    // then
+    final var valueCaptor = ArgumentCaptor.forClass(BufferWriter.class);
+    verify(logstream).valueWriter(valueCaptor.capture());
+    final var valueWriter = valueCaptor.getValue();
+    final var bytesWrittenToLogStream = new byte[valueWriter.getLength()];
+    final var valueBuffer = new UnsafeBuffer(bytesWrittenToLogStream);
+    valueWriter.write(valueBuffer, 0);
+    valueWriter.getLength();
+
+    assertThat(bytesWrittenToLogStream).isEqualTo(commandBytes);
+  }
+
+  private TopologyPartitionListenerImpl mockTopologyListener(
+      final int partitionId, final int brokerId) {
+    final var topologyListener = mock(TopologyPartitionListenerImpl.class);
+
+    final var topology = new Int2IntHashMap(-1);
+    topology.put(partitionId, brokerId);
+    when(topologyListener.getPartitionLeaders()).thenReturn(topology);
+
+    return topologyListener;
+  }
+
+  private Consumer<byte[]> getHandlerForPartition(
+      final ClusterCommunicationService communicationService, final int partitionId) {
+    @SuppressWarnings("unchecked")
+    final ArgumentCaptor<BiConsumer<MemberId, byte[]>> handlerCaptor =
+        ArgumentCaptor.forClass(BiConsumer.class);
+    final var executorCaptor = ArgumentCaptor.forClass(Executor.class);
+    verify(communicationService)
+        .subscribe(
+            eq(InterPartitionCommandSender.TOPIC_PREFIX + partitionId),
+            handlerCaptor.capture(),
+            executorCaptor.capture());
+    return (byte[] bytes) ->
+        executorCaptor
+            .getValue()
+            .execute(() -> handlerCaptor.getValue().accept(new MemberId("0"), bytes));
+  }
+
+  private byte[] sendCommand(
+      final Integer receiverBrokerId,
+      final Integer receiverPartitionId,
+      final ValueType valueType,
+      final Intent intent,
+      final byte[] command) {
+
+    final var topologyListener = mockTopologyListener(receiverPartitionId, receiverBrokerId);
+    final var sender = new InterPartitionCommandSenderImpl(communicationService, topologyListener);
+
+    final var buffer = new UnsafeBuffer(command);
+    final var bufferWriter = new DirectBufferWriter();
+    bufferWriter.wrap(buffer);
+    sender.sendCommand(receiverPartitionId, valueType, intent, bufferWriter);
+    scheduler.workUntilDone();
+
+    final var messageCaptor = ArgumentCaptor.forClass(byte[].class);
+    verify(communicationService)
+        .unicast(
+            eq(InterPartitionCommandSender.TOPIC_PREFIX + receiverPartitionId),
+            messageCaptor.capture(),
+            any());
+
+    return messageCaptor.getValue();
```

Comment:
:question: This would be same as the parameter `command`, right? So can just return it instead of using ArgumentCaptor?

Reply:
No, this is the entire message, including valueType and intent, not just the command. We are giving the sender the command and then capturing what the sender actually serialized and sends so that we can hand that over to the receiver.